### PR TITLE
[FLINK-13026][jdbc] Introduce JDBCLookupFunction

### PR DESCRIPTION
**Pull request by JingsongLi**
## What is the purpose of the change

It is very common to store dimension information in JDBC databases. It is very useful to support JDBCLookupFunction, user can use this to lookup dimension fields in his streaming job.
User can use JDBCLookupFunction to table function, or user can use JDBCLookupTableSource to blink planner temporal table join.

## Brief change log

1.Introduce JDBCLookupFunction
2.Introduce JDBCLookupTableSource

## Verifying this change

ut

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): (no)
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: (no)
  - The serializers: (no)
  - The runtime per-record code paths (performance sensitive): (no)
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Yarn/Mesos, ZooKeeper: (no)
  - The S3 file system connector: (no)

## Documentation

  - Does this pull request introduce a new feature? (yes)
  - If yes, how is the feature documented? (JavaDocs)

### CODE REVIEW

**Comment by wuchong** [comment]
Hi @JingsongLi , #8867 is merged, you can rebase this commit on that.

File: flink-connectors/flink-jdbc/src/main/java/org/apache/flink/api/java/io/jdbc/JDBCLookupFunction.java [hunk 1]
```diff
@@ -0,0 +1,340 @@
...
+ * The query template like:
+ * <PRE>
+ *   SELECT c, d, e, f from T where a = ? and b = ?
+ * </PRE>
+ *
+ * Support cache the result to avoid frequent accessing to remote databases.
+ * 1.The cacheMaxSize is -1 means not use cache.
+ * 2.For real-time data, you need to set the TTL of cache.
+ */
+public class JDBCLookupFunction extends TableFunction<Row> {
```
[the remark this answers is not in the document]
Comment (JingsongLi) [reply flink-connectors/flink-jdbc/src/main/java/org/apache/flink/api/java/io/jdbc/JDBCLookupFunction.java]: OK

File: flink-connectors/flink-jdbc/src/main/java/org/apache/flink/api/java/io/jdbc/JDBCLookupFunction.java [hunk 2]
```diff
@@ -0,0 +1,340 @@
...
+		this.outputSqlTypes = Arrays.stream(fieldTypes).mapToInt(JDBCTypeUtil::typeInformationToSqlType).toArray();
+
+		String quoteTableName = leftQuote + tableName + rightQuote;
+		String selectFields = StringUtils.join(Arrays.stream(fieldNames)
+				.map(name -> leftQuote + name + rightQuote)
+				.toArray(String[]::new), ",");
+		String conditions = StringUtils.join(Arrays.stream(keyNames)
+				.map(name -> leftQuote + name + rightQuote + " = ?")
+				.toArray(String[]::new), " AND ");
+		this.query = "SELECT " + selectFields + " FROM " + quoteTableName + " WHERE " + conditions;
```
[the remark this answers is not in the document]
Comment (JingsongLi) [reply flink-connectors/flink-jdbc/src/main/java/org/apache/flink/api/java/io/jdbc/JDBCLookupFunction.java]: OK

File: flink-connectors/flink-jdbc/src/main/java/org/apache/flink/api/java/io/jdbc/JDBCLookupFunction.java [hunk 3]
```diff
@@ -0,0 +1,340 @@
...
+		Class.forName(drivername);
+		if (username == null) {
+			dbConn = DriverManager.getConnection(dbURL);
+		} else {
+			dbConn = DriverManager.getConnection(dbURL, username, password);
+		}
+	}
+
+	@Override
+	public void close() throws IOException {
```
[the remark this answers is not in the document]
Comment (JingsongLi) [reply flink-connectors/flink-jdbc/src/main/java/org/apache/flink/api/java/io/jdbc/JDBCLookupFunction.java]: OK

**Comment by JingsongLi** [comment]
@wuchong Thanks review~ Code updated.

File: flink-connectors/flink-jdbc/src/main/java/org/apache/flink/api/java/io/jdbc/JDBCUtils.java [hunk 4]
```diff
@@ -150,4 +150,67 @@ public static void setField(PreparedStatement upload, int type, Row row, int ind
...
+				return set.getTime(index + 1);
+			case java.sql.Types.TIMESTAMP:
+				return set.getTimestamp(index + 1);
+			case java.sql.Types.BINARY:
+			case java.sql.Types.VARBINARY:
+			case java.sql.Types.LONGVARBINARY:
+				return set.getBytes(index + 1);
+			default:
+				LOG.warn("Unmanaged sql type ({}) for column {}. Best effort approach to get its value: {}.",
+						type, index + 1);
```
Comment (wuchong) [inline flink-connectors/flink-jdbc/src/main/java/org/apache/flink/api/java/io/jdbc/JDBCUtils.java]: value missed.

File: flink-connectors/flink-jdbc/src/main/java/org/apache/flink/api/java/io/jdbc/JDBCLookupOptions.java [hunk 5]
```diff
@@ -0,0 +1,83 @@
...
+	private final long cacheExpireMs;
+	private final int maxRetryTimes;
+
+	private JDBCLookupOptions(long cacheMaxSize, long cacheExpireMs, int maxRetryTimes) {
+		this.cacheMaxSize = cacheMaxSize;
+		this.cacheExpireMs = cacheExpireMs;
+		this.maxRetryTimes = maxRetryTimes;
+	}
+
+	public long getCacheMaxSize() {
```
Comment (wuchong) [inline flink-connectors/flink-jdbc/src/main/java/org/apache/flink/api/java/io/jdbc/JDBCLookupOptions.java]: Could you add some javadoc to the methods? What size and cache mean? 

The following method is the same.

File: flink-connectors/flink-jdbc/src/main/java/org/apache/flink/api/java/io/jdbc/JDBCTableSource.java [hunk 6]
```diff
@@ -0,0 +1,115 @@
...
+import org.apache.flink.table.sources.TableSource;
+import org.apache.flink.types.Row;
+
+import static org.apache.flink.util.Preconditions.checkNotNull;
+
+/**
+ * {@link TableSource} for JDBC.
+ * Now only support {@link LookupableTableSource}.
+ */
+public class JDBCTableSource implements LookupableTableSource<Row> {
```
Comment (wuchong) [inline flink-connectors/flink-jdbc/src/main/java/org/apache/flink/api/java/io/jdbc/JDBCTableSource.java]: This should also implement `StreamTableSource` in order to bridge to `TableSourceFactory`. We can return a bounded `StreamTableSource` using `JDBCInputFormat`.

**Comment by JingsongLi** [comment]
> Thanks @JingsongLi , overall looks good to me.
> 
> However, I find that we have so many `Builder`s and not all the methods of the Builders need to be called. Can we add some javadoc to them, which are mandatory and which are optional.

Yeah, I will add comments to all method. (required and optional)

File: flink-connectors/flink-jdbc/src/main/java/org/apache/flink/api/java/io/jdbc/JDBCUtils.java (hunk 4 above)
Comment (JingsongLi) [reply flink-connectors/flink-jdbc/src/main/java/org/apache/flink/api/java/io/jdbc/JDBCUtils.java]: OK

File: flink-connectors/flink-jdbc/src/main/java/org/apache/flink/api/java/io/jdbc/JDBCLookupOptions.java (hunk 5 above)
Comment (JingsongLi) [reply flink-connectors/flink-jdbc/src/main/java/org/apache/flink/api/java/io/jdbc/JDBCLookupOptions.java]: Yes

File: flink-connectors/flink-jdbc/src/main/java/org/apache/flink/api/java/io/jdbc/JDBCTableSource.java (hunk 6 above)
Comment (JingsongLi) [reply flink-connectors/flink-jdbc/src/main/java/org/apache/flink/api/java/io/jdbc/JDBCTableSource.java]: I think we can add it in next PR, and the `JDBCInputFormat` should be enhance to support filters.

**Comment by wuchong** [comment]
LGTM. 

@KurtYoung  do you need to have a look?

**Comment by zentol** [comment]
CI report for commit e99b2c60ff132995254505ac565cca31c44af3be: FAILURE [Build](https://travis-ci.org/flink-ci/flink-ci/builds/554872297?utm_source=github_status&utm_medium=notification)

**Pull request merged**